react/ProductContextProvider.tsx

```diff
@@ -4,6 +4,76 @@ import ProductContext from './ProductContext'
 import { ProductDispatchContext } from './ProductDispatchContext'
 import { useProductReducer, getSelectedItem } from './reducer'
 import { getSelectedSKUFromQueryString } from './modules/skuQueryString'
+import { MaybeProduct, Item } from './ProductTypes'
+
+export interface ProductAndQuery {
+  query: Record<string, any>
+  product: MaybeProduct
+}
+
+type GroupId = string
+
+export interface AssemblyOptionItem {
+  id: string
+  quantity: number
+  seller: string
+  initialQuantity: number
+  choiceType: string
+  name: string
+  price: number
+  children: Record<string, AssemblyOptionItem[]> | null
+}
+
+export interface BuyButtonContextState {
+  clicked: boolean
+}
+
+export interface ProductContextState {
+  selectedItem?: Item | null
+  product: MaybeProduct
+  selectedQuantity: number
+  skuSelector: {
+    selectedImageVariationSKU: string | null
+    isVisible: boolean
+    areAllVariationsSelected: boolean
+  }
+  buyButton: BuyButtonContextState
+  assemblyOptions: {
+    items: Record<GroupId, AssemblyOptionItem[]>
+    inputValues: Record<GroupId, InputValues>
+    areGroupsValid: Record<GroupId, boolean>
+  }
```

Comment:
I don't know about these ones, but could you confirm that these are never `undefined` or `null`?

Reply:
Yep! They could be empty objects though. But I guess the Record type would already cover that, right?

Reply:
The `Record` can only be an empty object, but it doesn't cover if `assemblyOptions.items === undefined` for instance

Reply:
Is that ok?

Reply:
Yeep, no problem